### PR TITLE
Added optional filter params for groups datasource

### PR DESCRIPTION
**Pull request by vmanilo**
Resolves (github issue):
https://github.com/Twingate/terraform-provider-twingate/issues/137
https://github.com/Twingate/terraform-provider-twingate/issues/99
https://github.com/Twingate/terraform-provider-twingate/issues/139

## Changes
- fixed groups datasource

### CODE REVIEW

File: twingate/datasource_groups.go [hunk 1]
```diff
@@ -22,30 +28,74 @@ func datasourceGroupsRead(ctx context.Context, resourceData *schema.ResourceData
...
+			"is_active": bool(group.IsActive),
 		})
 	}
 
 	if err := resourceData.Set("groups", data); err != nil {
 		return diag.FromErr(err)
 	}
 
-	resourceData.SetId(groupName)
+	resourceData.SetId("groups-by-filter")
```
Comment (tjstansell) [inline twingate/datasource_groups.go:42]: I don't know enough about how datasources work, but does setting this to a static value:
* cause refresh problems?
* cause problems with multiple data sources?

File: docs/data-sources/groups.md [hunk 2]
```diff
@@ -24,9 +24,11 @@ data "twingate_groups" "foo" {
 <!-- schema generated by tfplugindocs -->
 ## Schema
 
-### Required
+### Optional
 
+- `is_active` (Boolean) Indicates if the Group is active
```
Comment (tjstansell) [inline docs/data-sources/groups.md:29]: Personally, I would use descriptions that indicate exactly how these are used to filter.
```
- `is_active` (Boolean) Returns only groups matching the specified state.
- `name` (String) Returns only groups that exactly match this name.
- `type` (String) Returns only groups of the specified type (valid: `MANUAL`, `SYNCED`, `SYSTEM`).
```

File: twingate/client_group.go [hunk 3]
```diff
@@ -210,3 +230,65 @@ func (client *Client) deleteGroup(ctx context.Context, groupID graphql.ID) error
...
+
+	if f.IsActive != nil && *f.IsActive != bool(group.IsActive) {
+		return false
+	}
+
+	return true
+}
+
+func (client *Client) filterGroups(ctx context.Context, filter *GroupsFilter) ([]*Group, error) {
+	if filter == nil {
```
Comment (tjstansell) [inline twingate/client_group.go:253]: If `buildFilter()` always returned a valid `GroupsFilter` object (even if all the attributes are `nil`), you could remove this whole block to simplify the code a bit.  You're already optimizing below which query you want to make to the API based on the individual filters, so this block is essentially a duplicate of all that.

Comment (vmanilo) [reply twingate/client_group.go:253]: thank you, good point

File: twingate/datasource_groups.go (hunk 1 above)
Comment (vmanilo) [reply twingate/datasource_groups.go:42]: tested locally -  all good

also added test with two datasource

File: docs/data-sources/groups.md (hunk 2 above)
Comment (vmanilo) [reply docs/data-sources/groups.md:29]: @alexmensch what do you think?

File: twingate/client_group.go (hunk 3 above)
Comment (vmanilo) [reply twingate/client_group.go]: fixed

File: docs/data-sources/groups.md (hunk 2 above)
Comment (alexmensch) [reply docs/data-sources/groups.md:29]: Yes agreed -- @vmanilo go ahead as above, just need to capitalize groups to `Groups`. Thanks all!

**Review by alexmensch — CHANGES_REQUESTED** [review]

File: docs/data-sources/groups.md (hunk 2 above)
Comment (vmanilo) [reply docs/data-sources/groups.md]: fixed

**Review by alexmensch — APPROVED** [review]

File: twingate/datasource_groups.go [hunk 4]
```diff
@@ -22,29 +29,73 @@ func datasourceGroupsRead(ctx context.Context, resourceData *schema.ResourceData
...
 
+func buildFilter(resourceData *schema.ResourceData) *GroupsFilter {
+	groupName, hasName := resourceData.GetOk("name")
+	groupType, hasType := resourceData.GetOk("type")
+
+	// GetOk does not provide correct value for exists flag (second output value)
+	groupIsActive, hasIsActive := resourceData.GetOkExists("is_active") // nolint:staticcheck
+
+	if !hasName && !hasType && !hasIsActive {
+		return nil
```
Comment (tjstansell) [inline twingate/datasource_groups.go:56]: How is this not causing problems when there is no filtering and you pull all groups?

Comment (vmanilo) [reply twingate/datasource_groups.go:56]: I'm handling nil case in the GroupsFilter methods

File: twingate/client_group.go [hunk 5]
```diff
@@ -214,3 +233,60 @@ func (client *Client) deleteGroup(ctx context.Context, groupID graphql.ID) error
...
+			return nil, nil
+		}
+
+		return nil, err
+	}
+
+	var filtered []*Group
+
+	for _, g := range groups {
+		if filter.Match(g) {
```
Comment (ekampf) [inline twingate/client_group.go:286]: @alexmensch did we agree not to allow filtering by `name` until server api supports it?

File: twingate/datasource_groups_test.go [hunk 6]
```diff
@@ -84,19 +86,161 @@ func testTwingateGroupsDoesNotExists(name string) string {
...
+		resource.Test(t, resource.TestCase{
+			ProviderFactories: testAccProviderFactories,
+			PreCheck:          func() { testAccPreCheck(t) },
+			CheckDestroy:      testAccCheckTwingateGroupDestroy,
+			Steps: []resource.TestStep{
+				{
+					Config: testDatasourceTwingateGroupsWithDatasource(groupName),
+					Check: resource.ComposeTestCheckFunc(
+						resource.TestCheckOutput("my_group", groupName),
+						resource.TestCheckResourceAttr("data.twingate_groups.out", "groups.#", "2"),
```
Comment (ekampf) [inline twingate/datasource_groups_test.go:205]: Fix test - if both DS have the same value how do we know they're not overriding each other?

Comment (vmanilo) [reply twingate/datasource_groups_test.go]: updated test

File: twingate/client_group.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexmensch) [reply twingate/client_group.go:286]: Yes that's correct

[the remark this answers is not in the document]
Comment (alexmensch) [reply twingate/client_group.go:286]: But if it's implemented already...may as well ship it? What do you think @ekampf?

**Pull request merged**